sphinx/locale/__init__.py

```diff
@@ -188,119 +198,152 @@ def lazy_gettext(string):
     return _TranslationProxy(mygettext, string)  # type: ignore
 
 
-l_ = lazy_gettext
+translators = defaultdict(NullTranslations)  # type: Dict[Tuple[unicode, unicode], NullTranslations]  # NOQA
 
 
-admonitionlabels = {
-    'attention': l_('Attention'),
-    'caution':   l_('Caution'),
-    'danger':    l_('Danger'),
-    'error':     l_('Error'),
-    'hint':      l_('Hint'),
-    'important': l_('Important'),
-    'note':      l_('Note'),
-    'seealso':   l_('See also'),
-    'tip':       l_('Tip'),
-    'warning':   l_('Warning'),
-}  # type: Dict[unicode, unicode]
-
-versionlabels = {
-    'versionadded':   l_('New in version %s'),
-    'versionchanged': l_('Changed in version %s'),
-    'deprecated':     l_('Deprecated since version %s'),
-}  # type: Dict[unicode, unicode]
-
-# XXX Python specific
-pairindextypes = {
-    'module':    l_('module'),
-    'keyword':   l_('keyword'),
-    'operator':  l_('operator'),
-    'object':    l_('object'),
-    'exception': l_('exception'),
-    'statement': l_('statement'),
-    'builtin':   l_('built-in function'),
-}  # Dict[unicode, _TranslationProxy]
-
-translators = {}  # type: Dict[unicode, Any]
-
-if PY3:
-    def _(message, *args):
-        # type: (unicode, *Any) -> unicode
-        try:
-            if len(args) <= 1:
-                return translators['sphinx'].gettext(message)
-            else:  # support pluralization
-                return translators['sphinx'].ngettext(message, args[0], args[1])
-        except KeyError:
-            return message
-else:
-    def _(message, *args):
-        # type: (unicode, *Any) -> unicode
-        try:
-            if len(args) <= 1:
-                return translators['sphinx'].ugettext(message)
-            else:  # support pluralization
-                return translators['sphinx'].ungettext(message, args[0], args[1])
-        except KeyError:
-            return message
-
-
-def __(message, *args):
-    # type: (unicode, *Any) -> unicode
-    """A dummy wrapper to i18n'ize exceptions and command line messages.
-
-    In future, the messages are translated using LC_MESSAGES or any other
-    locale settings.
-    """
-    return message if len(args) <= 1 else args[0]
-
-
-def init(locale_dirs, language, catalog='sphinx'):
+def init(locale_dirs, language, catalog='sphinx', namespace='general'):
     # type: (List, unicode, unicode) -> Tuple[Any, bool]
     """Look for message catalogs in `locale_dirs` and *ensure* that there is at
     least a NullTranslations catalog set in `translators`.  If called multiple
     times or if several ``.mo`` files are found, their contents are merged
     together (thus making ``init`` reentrable).
     """
     global translators
-    translator = translators.get(catalog)
+    translator = translators.get((namespace, catalog))
     # ignore previously failed attempts to find message catalogs
-    if isinstance(translator, gettext.NullTranslations):
+    if isinstance(translator, NullTranslations):
```

Comment:
It seems this merging process does not work.
All translations object is a subclass of `NullTranslations`!
```
>>> import gettext
>>> isinstance(gettext.GNUTranslations(), gettext.NullTranslations)
True
```